### DISCUSSION TITLE
Query JSON with XPath

### DISCUSSION THREAD

**Discussion by arenas-guerrero-julian** [morph-kgc/morph-kgc#137] [Ideas]
Last time I checked I could not find an implementation of JSONPath in python that was comprehensive and actively maintained. However, for XPath there are a couple of very good implementations.

The [jsonpath library](https://pypistats.org/packages/jsonpath-python) currently used by Morph-KGC has limitations and some bugs (e.g., #132). I though of internally converting JSON to XML (e.g., using [json2xml](https://github.com/vinitkumar/json2xml#omit-list-item)) and using XPath to query JSON.

Some advantages:
- The XPath implementations available in python are good.
- XPath is standardized, JSONPath is not.
- XPath provides way more functions than JSONPath.
- This remains compliant w.r.t. RML (JSONPath is the default reference formulation for JSON, but RML does not enforce this language for JSON).

Some disadvantages:
- Conversion to XML may add some overhead in materialization.
- Mapping developers are used to using JSONPath for JSON.

@dgarijo @ocorcho @anaigmo what do you think? Do you see any other drawback? I am not that familiar with hierarchical data. A possible solution in the middle is to allow both for JSON, JSONPath and XPath, and add a configuration option to select which one to use.

**Comment by dgarijo** [comment]
As a user, I don't mind waiting a little more.
I do mind, however, if I wrap up my mappings in YARRRML and I don't get them to work with morph-KGC. My suggestion is to add whatever that makes my workflow easier, i.e., I need to be able to test my mappings in YARRRML without going to RML.

**Reply by arenas-guerrero-julian** [reply]
Translating YARRRML to RML and then executing Morph-KGC should work fine. However, you could not test your mappings in matey and then use the generated RML mapping with Morph-KGC.

This idea could wait for the moment in which Morph-KGC natively supports YARRRML so that Matey is not necessary for and testing the mappings can be done directly in Morph-KGC.

**Reply by dgarijo** [reply]
Then I would suggest waiting until having YARRRML support. At least I would not test new XPath configurations directly in RML, it's too cumbersome :(

**Comment by arenas-guerrero-julian** [comment]
As far as I could read [here](https://docs.marklogic.com/guide/app-dev/json) @marklogic follows this strategy querying JSON with XPath. I still need to read this better. Because we are suportting YARRRML in the **short term** I think there would be no problem with this @dgarijo.

**Reply by arenas-guerrero-julian** [reply]
[JSON-to-XML](https://www.w3.org/TR/xslt-30/#json).

**Comment by arenas-guerrero-julian** [comment]
Another possibility is to try this library https://github.com/besok/jsonpath-rust/issues/29
